### DISCUSSION TITLE
Guidance

### DISCUSSION THREAD

**Discussion by richb201** [BretFisher/ama#133] [Q&A]
I am not a docker expert and don't intend to become one at this late date in my career. I have a code igniter web app which talks with a Chrome extension. It runs on aws lightsail in a Bitnami lamp environment. I wish to be able to develop and test enhancements to it on my ubuntu laptop.

Bitnami has pre made docker images, one for php, more for MariaDb, one for apache, and one for code igniter. I managed to docker-compose up and all seems aok. I also use grocerrycrud which is an add on for Code igniter. I need to bring my application with all its screens and php files and "inject" them along with the CI config files into the code igniter created structure in the running container. Locations are important. 

I was thinking I could use the Bitnami pre made images and then layer my code on top in docker-compose, but isn't layering something that gets done in a Dockerfile instead?

There are so many options but I need to pick one that has a reasonable chance of working. Thanks and I look forward to reading your answer. Btw, will I get a link to the answer in email?

**Comment by BretFisher** [comment]
First, understand I don't know anything about those images or your code :)

How your images are used for creating containers on servers vs. local dev use are different.

You want your prod images to container your app code, so you would have a Dockerfile that uses COPY to bundle everything in the image you need for production use of the container. You'd store that image in a registry and your server would pull that image to start a container.

In local dev/test, you usually want to bind-mount your host copy of your code into the container at runtime. You'd use docker-compose for that, and bind-mount it in using the docker-compose.yml. That YAML file can also be stored in your code repo with the Dockerfile.

Does that  help?

**Comment by richb201** [comment]
Yes, that makes sense. I surely get the prod images part. I am confused about bind mounting.  Why at runtime? I figured that I would need to do the COPY of my code into the development image so I could debug it. What container dir do I copy my source code into? The examples always show /app, no idea why?

Thks

On September 23, 2019, at 13:50, Bret Fisher <notifications@github.com> wrote:

First, understand I don't know anything about those images or your code :)

How your images are used for creating containers on servers vs. local dev use are different.

You want your prod images to container your app code, so you would have a Dockerfile that uses COPY to bundle everything in the image you need for production use of the container. You'd store that image in a registry and your server would pull that image to start a container.

In local dev/test, you usually want to bind-mount your host copy of your code into the container at runtime. You'd use docker-compose for that, and bind-mount it in using the docker-compose.yml. That YAML file can also be stored in your code repo with the Dockerfile.

Does that help?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.￼

[ { "@context": "<a href="http://schema.org">http://schema.org</a>", "@type": "EmailMessage", "potentialAction": { "@type": "ViewAction", "target": "<a href="https://github.com/BretFisher/ama/issues/69?email_source=notifications">https://github.com/BretFisher/ama/issues/69?email_source=notifications</a>\u0026email_token=ABVBHGJAQE7FQ44KHUSUACLQLD6VHA5CNFSM4IYBU47KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7LWK5I#issuecomment-534209909", "url": "<a href="https://github.com/BretFisher/ama/issues/69?email_source=notifications">https://github.com/BretFisher/ama/issues/69?email_source=notifications</a>\u0026email_token=ABVBHGJAQE7FQ44KHUSUACLQLD6VHA5CNFSM4IYBU47KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7LWK5I#issuecomment-534209909", "name": "View Issue" }, "description": "View this Issue on GitHub", "publisher": { "@type": "Organization", "name": "GitHub", "url": "<a href="https://github.com">https://github.com</a>" } } ]

**Comment by BretFisher** [comment]
Mounting a volume (named volume, volume mount, etc.) is always done at
runtime. It's a host-specific thing (the host has to have those files). Yes
you would want a COPY in your Dockerfile so your image has your app, but
when you bind-mount over top of that at runtime, the bind-mount wins. The
location doesn't matter as long as it isn't a system path, so common ones
like /app, /srv/src/, /var/code, etc.
